cli/azd/pkg/project/service_manager.go

```diff
@@ -0,0 +1,465 @@
+
+	Restore(
+		ctx context.Context,
+		serviceConfig *ServiceConfig,
+	) *async.TaskWithProgress[*ServiceRestoreResult, ServiceProgress]
+	Build(
+		ctx context.Context,
+		serviceConfig *ServiceConfig,
+		restoreOutput *ServiceRestoreResult,
+	) *async.TaskWithProgress[*ServiceBuildResult, ServiceProgress]
+	Package(
+		ctx context.Context,
+		serviceConfig *ServiceConfig,
+		buildOutput *ServiceBuildResult,
+	) *async.TaskWithProgress[*ServicePackageResult, ServiceProgress]
+	Publish(
+		ctx context.Context,
+		serviceConfig *ServiceConfig,
+		packageOutput *ServicePackageResult,
+	) *async.TaskWithProgress[*ServicePublishResult, ServiceProgress]
+	Deploy(
+		ctx context.Context,
+		serviceConfig *ServiceConfig,
+	) *async.TaskWithProgress[*ServiceDeployResult, ServiceProgress]
+
+	GetFrameworkService(ctx context.Context, serviceConfig *ServiceConfig) (FrameworkService, error)
+	GetServiceTarget(ctx context.Context, serviceConfig *ServiceConfig) (ServiceTarget, error)
+}
+
+type serviceManager struct {
+	azdContext      *azdcontext.AzdContext
+	env             *environment.Environment
+	commandRunner   exec.CommandRunner
+	azCli           azcli.AzCli
+	console         input.Console
+	accountManager  account.Manager
+	resourceManager ResourceManager
+	serviceLocator  ioc.ServiceLocator
+}
+
+func NewServiceManager(
+	azdContext *azdcontext.AzdContext,
+	env *environment.Environment,
+	commandRunner exec.CommandRunner,
+	azCli azcli.AzCli,
+	console input.Console,
+	accountManager account.Manager,
+	resourceManager ResourceManager,
+	serviceLocator ioc.ServiceLocator,
+) ServiceManager {
+	return &serviceManager{
+		azdContext:      azdContext,
+		env:             env,
+		commandRunner:   commandRunner,
+		azCli:           azCli,
+		console:         console,
+		accountManager:  accountManager,
+		resourceManager: resourceManager,
+		serviceLocator:  serviceLocator,
+	}
+}
+
+// Gets a list of required tools for the current service
+func (sm *serviceManager) GetRequiredTools(ctx context.Context, serviceConfig *ServiceConfig) ([]tools.ExternalTool, error) {
+	frameworkService, err := sm.GetFrameworkService(ctx, serviceConfig)
+	if err != nil {
+		return nil, fmt.Errorf("getting framework services: %w", err)
+	}
+
+	serviceTarget, err := sm.GetServiceTarget(ctx, serviceConfig)
+	if err != nil {
+		return nil, fmt.Errorf("getting service target: %w", err)
+	}
+
+	requiredTools := []tools.ExternalTool{}
+	requiredTools = append(requiredTools, frameworkService.RequiredExternalTools(ctx)...)
+	requiredTools = append(requiredTools, serviceTarget.RequiredExternalTools(ctx)...)
+
+	return requiredTools, nil
+}
+
+func (sm *serviceManager) Restore(
+	ctx context.Context,
+	serviceConfig *ServiceConfig,
+) *async.TaskWithProgress[*ServiceRestoreResult, ServiceProgress] {
+	return async.RunTaskWithProgress(func(task *async.TaskContextWithProgress[*ServiceRestoreResult, ServiceProgress]) {
```

Comment:
As you may have noticed, once we start using `async`, the entire structure of the program now necessarily needs to use the `TaskContext` to return or set values and ends up clashing with writing `go` code that simply expresses results using `return` statements. And the code becomes `async` all the way down. I personally think that the added boilerplate isn't worth it, as `async` without language support is tedious.

I would love for us to have this discussion as a team, before we start introducing this. I would love we could make this decision outside of the refactoring PR.

Reply:
Yes, this observation is correct.  As soon as you introduce async behavior any consumers of this also need to become and respect this.